src/thirty_two.rs

```diff
@@ -0,0 +1,318 @@
+        if self.buffer_usage + bytes.len() < self.buffer.len() {
+            unsafe {
+                let tail = self.buffer.as_mut_ptr().offset(self.buffer_usage as isize);
+                std::ptr::copy_nonoverlapping(bytes.as_ptr(), tail, bytes.len());
+            }
+            self.buffer_usage += bytes.len();
+            return;
+        }
+
+        // Some data left from previous update. Fill the buffer and
+        // consume it first.
+        if self.buffer_usage > 0 {
+            let bytes_to_use = self.buffer.len() - self.buffer_usage;
+            let (to_use, leftover) = bytes.split_at(bytes_to_use);
+
+            unsafe {
+                let tail = self.buffer.as_mut_ptr().offset(self.buffer_usage as isize);
+                std::ptr::copy_nonoverlapping(to_use.as_ptr(), tail, bytes_to_use);
+            }
+
+            let (iter, _) = self.buffer.u32_stream();
+
+            self.core.ingest_chunks(iter);
+
+            bytes = leftover;
+            self.buffer_usage = 0;
+        }
+
+        // Consume the input data in large chunks
+        let (iter, bytes) = bytes.u32_stream_with_stride(4);
+        self.core.ingest_chunks(iter);
+
+        // Save any leftover data for the next call
+        if bytes.len() > 0 {
+            unsafe {
+                std::ptr::copy_nonoverlapping(bytes.as_ptr(), self.buffer.as_mut_ptr(), bytes.len());
+            }
+            self.buffer_usage = bytes.len();
+        }
+    }
+
+    fn finish(&self) -> u64 { // NODIFF
+        let mut hash;
+
+        // We have processed at least one full chunk
+        if self.total_len >= CHUNK_SIZE as u32 {
+            hash = self.core.finish();
+        } else {
+            hash = self.seed.wrapping_add(PRIME_5);
+        }
+
+        hash = hash.wrapping_add(self.total_len);
+
+        let buffered = &self.buffer[..self.buffer_usage];
+        let (buffered_u32s, buffered) = buffered.u32_stream();
+
+        for k1 in buffered_u32s {
+            let k1 = k1.wrapping_mul(PRIME_3);
+            hash = hash.wrapping_add(k1);
+            hash = hash.rotate_left(17);
+            hash = hash.wrapping_mul(PRIME_4);
+        }
+
+        for buffered_u8 in buffered {
+            let k1 = (*buffered_u8 as u32).wrapping_mul(PRIME_5);
+            hash = hash.wrapping_add(k1);
+            hash = hash.rotate_left(11);
+            hash = hash.wrapping_mul(PRIME_1);
+        }
+
+        // The final intermixing
+        hash ^= hash >> 15;
+        hash = hash.wrapping_mul(PRIME_2);
+        hash ^= hash >> 13;
+        hash = hash.wrapping_mul(PRIME_3);
+        hash ^= hash >> 16;
+
+        hash as u64
+    }
+}
+
+#[derive(Clone)]
+/// Constructs a randomized seed and reuses it for multiple hasher instances. See the usage warning on `XxHash32`.
+pub struct RandomXxHashBuilder(u32);
+
+impl RandomXxHashBuilder {
+    fn new() -> RandomXxHashBuilder {
+        RandomXxHashBuilder(rand::thread_rng().gen())
```

Comment:
There is a risk to exhaust the OS random pool, you could refer to [RandomState](https://doc.rust-lang.org/src/std/collections/hash/map.rs.html#2128), or use a PRN seed like [Xoroshiro128Rng](https://github.com/flier/rust-fasthash/blob/master/fasthash/src/hasher.rs#L143).